### PR TITLE
[Queue Mode][RSpec] Pass each batch of tests to the queue hooks: `KnapsackPro::Hooks::Queue.before_subset_queue` and `KnapsackPro::Hooks::Queue.after_subset_queue`

### PR DESCRIPTION
**Pull request by ArturT**
# Story

[Link to the internal story](https://trello.com/c/KJ9f5hSP)

## Related

Inspired by draft of PR: 

* https://github.com/KnapsackPro/knapsack_pro-ruby/pull/252

# Description

In RSpec when you use Queue Mode, the queue hooks `KnapsackPro::Hooks::Queue.before_subset_queue` and `KnapsackPro::Hooks::Queue.after_subset_queue` get a 3rd variable - the `queue`. 

  The `queue` variable stores an enumerable collection with each batch of tests fetched from the Queue API. The batch has:

  * a list of test file paths (`KnapsackPro::Batch#test_file_paths` returns an array like `['a_spec.rb', 'b_spec.rb']`)
  * a status of the given set of tests in the batch (`KnapsackPro::Batch#status` returns `:not_executed`, `:passed` or `:failed`)

Example usage:

```ruby
# spec_helper.rb

KnapsackPro::Hooks::Queue.before_subset_queue do |queue_id, subset_queue_id, queue|
  print "Tests from all batches fetched from the Queue API so far:  "
  puts queue.map(&:test_file_paths).inspect

  queue.each(&:test_file_paths) # you can use each as well

  print "Current batch tests: "
  puts queue.current_batch.test_file_paths.inspect

  print "Current batch status: "
  puts queue.current_batch.status # returns :not_executed in the `before_subset_queue` hook
end

KnapsackPro::Hooks::Queue.after_subset_queue do |queue_id, subset_queue_id, queue|
  print "Tests from all batches fetched from the Queue API so far: "
  puts queue.map(&:test_file_paths).inspect

  print "Current batch tests: "
  puts queue.current_batch.test_file_paths.inspect

  print "Current batch status: "
  puts queue.current_batch.status # returns :passed or :failed in the `after_subset_queue` hook
end
```

# Checklist reminder

- [x] You follow the architecture outlined below for RSpec in Queue Mode, which is a work in progress (feel free to propose changes):
  - Pure: `lib/knapsack_pro/pure/queue/rspec_pure.rb` contains pure functions that are unit tested.
  - Extension: `lib/knapsack_pro/extensions/rspec_extension.rb` encapsulates calls to RSpec internals and is integration and e2e tested.
  - Runner: `lib/knapsack_pro/runners/queue/rspec_runner.rb` invokes the pure code and the extension to produce side effects, which are integration and e2e tested.

### CODE REVIEW

File: lib/knapsack_pro/queue.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    def each(&block)
+      @batches.each(&block)
+    end
+
+    def map(&block)
+      @batches.map(&block)
+    end
+
```
Comment (3v0k4) [inline lib/knapsack_pro/queue.rb:18]: By implementing `Enumerable`, you get `map` for free:

```suggestion
```

File: lib/knapsack_pro/batch.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    def _passed
+      @passed = true
+    end
+
+    def _failed
+      @passed = false
+    end
+  end
```
Comment (3v0k4) [inline lib/knapsack_pro/batch.rb:30]: I'd propose:

```suggestion
  class Batch
    attr_reader :test_file_paths

    def initialize(test_file_paths)
      @test_file_paths = test_file_paths
      @status = :not_executed
    end

    def executed?
      @status != :not_executed
    end

    def passed?
      @status == :passed
    end

    def _passed
      @status = :passed
    end

    def _failed
      @status = :failed
    end
  end
```

File: spec/integration/runners/queue/rspec_runner_spec.rb [hunk 3]
```diff
@@ -739,6 +739,142 @@ def when_first_matching_example_defined(type:)
...
+      expect(actual.stdout).to include('Executed batches in after_subset_queue: [true, true, true]')
+      expect(actual.stdout).to include('Executed batches in after_subset_queue: [true, true, true, true]')
+
+      expect(actual.stdout).to include('Current batch tests: ["spec_integration/a_spec.rb", "spec_integration/b_spec.rb"]')
+      expect(actual.stdout).to include('Current batch tests: ["spec_integration/c_spec.rb", "spec_integration/d_spec.rb"]')
+      expect(actual.stdout).to include('Current batch tests: ["spec_integration/e_spec.rb", "spec_integration/f_spec.rb"]')
+      expect(actual.stdout).to include('Current batch tests: ["spec_integration/g_spec.rb", "spec_integration/h_spec.rb"]')
+      expect(actual.stdout).to include('Current batch executed: true')
+      expect(actual.stdout).to include('Current batch passed: true')
+      expect(actual.stdout).to include('Current batch passed: false')
```
Comment (3v0k4) [inline spec/integration/runners/queue/rspec_runner_spec.rb:874]:
```suggestion
      expect(actual.stdout).to include('Current batch executed: true').exactly(4).times
      expect(actual.stdout).to include('Current batch passed: true').twice
      expect(actual.stdout).to include('Current batch passed: false').twice
```

File: lib/knapsack_pro/hooks/queue.rb [hunk 4]
```diff
@@ -54,22 +54,24 @@ def call_before_queue
           end
         end
 
-        def call_before_subset_queue
+        def call_before_subset_queue(queue = nil)
```
Comment (3v0k4) [inline lib/knapsack_pro/hooks/queue.rb:57]:
```suggestion
        # `queue` is always present for RSpec
        def call_before_subset_queue(queue = nil)
```

File: spec/knapsack_pro/queue_spec.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+describe KnapsackPro::Queue do
+  it 'simulates a Queue Mode flow' do
```
Comment (3v0k4) [inline spec/knapsack_pro/queue_spec.rb:2]:
```suggestion
  it 'simulates a Queue Mode build' do
```

File: spec/knapsack_pro/queue_spec.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    queue = described_class.new
+
+    expect(queue.current_batch).to be_nil
+
+    # 1st batch
+    batched_tests_1 = ['a_spec.rb', 'b_spec.rb']
```
Comment (3v0k4) [inline spec/knapsack_pro/queue_spec.rb:8]:
```suggestion
    test_files_paths_1 = ['a_spec.rb', 'b_spec.rb']
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,57 @@
 # Changelog
 
+### 7.3.0
+
+* [Queue Mode][RSpec] Add info about batch of tests in queue hooks: `KnapsackPro::Hooks::Queue.before_subset_queue` and `KnapsackPro::Hooks::Queue.after_subset_queue`
```
Comment (3v0k4) [inline CHANGELOG.md:5]:
```suggestion
* [Queue Mode][RSpec] Pass each batch of tests to the queue hooks: `KnapsackPro::Hooks::Queue.before_subset_queue` and `KnapsackPro::Hooks::Queue.after_subset_queue`
```

File: lib/knapsack_pro/batch.rb (hunk 2 above)
Comment (3v0k4) [reply lib/knapsack_pro/batch.rb:30]: Actually, I'd just expose `@status` and avoid `executed?` and `passed?`. I thinks it simplifies even more.

**Review by 3v0k4 — COMMENTED** [review]
Fantastic job!

If you end up applying all the comments related to the public API (`batch.rb`), consider this approved.

Happy to have a chat if you have any doubts about the CR.

File: lib/knapsack_pro/batch.rb (hunk 2 above)
Comment (ArturT) [reply lib/knapsack_pro/batch.rb]: Ok. I've applied changes.

**Comment by ArturT** [comment]
> Fantastic job!
> 
> If you end up applying all the comments related to the public API (`batch.rb`), consider this approved.
> 
> Happy to have a chat if you have any doubts about the CR.

@3v0k4 I applied all the comments. I'll release the new gem version today.

**Pull request merged**